README.rst
channels_rabbitmq
=================

A Django Channels channel layer that uses RabbitMQ as its backing store.

Does not support `Worker and Background Tasks
<https://channels.readthedocs.io/en/stable/topics/worker.html>`_.
(See `Rationale
<https://github.com/CJWorkbench/channels_rabbitmq/pull/11#issuecomment-499185070>`_
and use ``await get_channel_layer().current_connection`` to send to job queues.)

Works with Python 3.8, 3.9, 3.10

Installation
------------

``pip install channels_rabbitmq``

Usage
-----

Then set up the channel layer in your Django settings file like so::

    CHANNEL_LAYERS = {
        "default": {
            "BACKEND": "channels_rabbitmq.core.RabbitmqChannelLayer",
            "CONFIG": {
                "host": "amqp://guest:guest@127.0.0.1/asgi",
                # "ssl_context": ... (optional)
            },
        },
    }

Possible options for ``CONFIG`` are listed below.

``host``
~~~~~~~~

URL of the server to connect to, adhering to `RabbitMQ spec
<https://www.rabbitmq.com/uri-spec.html>`_. To connect to a RabbitMQ cluster,
use a DNS server to resolve a hostname to multiple IP addresses.
channels_rabbitmq will automatically reconnect if at least one of them is
reachable in case of a disconnection.

``expiry``
~~~~~~~~~~

Minimum number of seconds a message should wait in a RabbitMQ queue, before it
may be silently dropped.

Defaults to ``60``. You generally shouldn't need to change this, but you may
want to turn it down if you have peaky traffic you wish to drop, or up if you
have peaky traffic you want to backlog until you get to it.

``local_capacity``
~~~~~~~~~~~~~~~~~~

Number of incoming messages queued in memory. Defaults to ``100``. (A message
sent to a group with two channels counts as one message.) When ``local_capacity``
messages are queued, the message backlog will grow on RabbitMQ.

(This controls the ``prefetch_count`` on the RabbitMQ queue.)

``local_expiry``
~~~~~~~~~~~~~~~~

Minimum number of seconds a message received from RabbitMQ must be held in
memory waiting for ``receive()``, before it may be dropped. Defaults to
``expiry``.

A warning will be logged when a message expires locally. The warning can
indicate that a channel has more messages than it can handle; or that
messages are being sent to a channel that does not exist. (Perhaps a missing
channel was implied by ``group_add()``, and a matching ``group_discard()``
was never called.)

If ``local_expiry < expiry``, then you can end up ignoring (and logging)
messages locally while they still exist in the RabbitMQ queue. These messages
will be acked, so RabbitMQ will behave as though they were delivered.

``remote_capacity``
~~~~~~~~~~~~~~~~~~~

Number of messages stored on RabbitMQ for each client. Defaults to ``100``.
(A message sent to a group with three channels on two distinct clients counts
as two messages.) When ``remote_capacity`` messages are queued in RabbitMQ,
the channel will refuse new messages. Calls from any client to ``send()`` or
``group_send()`` to the at-capacity client will raise ``ChannelFull``.

``ssl_context``
~~~~~~~~~~~~~~~

An `SSL context
<https://docs.python.org/3/library/ssl.html#ssl-contexts>`_. Changes the
default ``host`` port to 5671 (instead of 5672).

For instance, to connect to an TLS RabbitMQ service that will verify your
client::

    import ssl
    ssl_context = ssl.create_default_context(
        cafile=str(Path(__file__).parent.parent / 'ssl' / 'server.cert'),
    )
    ssl_context.load_cert_chain(
        certfile=str(Path(__file__).parent.parent / 'ssl' / 'client.certchain'),
        keyfile=str(Path(__file__).parent.parent / 'ssl' / 'client.key'),
    )
    CHANNEL_LAYERS['default']['CONFIG']['ssl_context'] = ssl_context

By default, there is no SSL context; all messages (and passwords) are
are transmitted in cleartext.

``groups_exchange``
~~~~~~~~~~~~~~~~~~~

Global direct exchange name used by channels to exchange group messages.
Defaults to ``"groups"``. See also `Design decisions`_.

Accessing Carehare
------------------

We use `carehare`_ for its thorough handling of errors.

Django Channels' specification does not account for "connecting" and
"disconnecting". This layer does its best by constantly reconnecting, forever.

Call ``await get_channel_layer().current_connection`` to access an open Carehare
connection. This lets you use job queues without "Worker and Background Tasks".
Like this::

    # raise asyncio.CancelledError on failure
    connection = await get_channel_layer().carehare_connection

    # raise carehare.ConnectionClosed or carehare.ChannelClosed on error
    await connection.publish(b"task", routing_key="job_queue")

(The Carehare documentation explains how to build workers.)

A note on errors: a "connected" connection isn't guaranteed to *stay* connected
throughout every publish. It was merely connected *at some point in the past*.
When a disconnect occurs, all pending operations on that connection will raise
``carehare.ConnectionClosed``. This channel layer will log the error, and
``get_channel_layer().carehare_connection`` will point to a new Future. (This
error+reconnect is *guaranteed to happen* in production.)

Publish messages from Celery
----------------------------

Many `Celery
<https://docs.celeryproject.org/en/stable/getting-started/introduction.html>`_
users want to send messages to websockets users.

This is doable, though not intuitive. Don't use any Django Channels code:
Channels layers depend on long-running connections, and Celery bans those.
Same goes for carehare: don't use it from Celery.

Instead, from a Celery-worker ``@task`` you can send messages to your
Django-Channels consumers using Celery's RabbitMQ connection::

    from typing import Any, Dict

    import msgpack

    def publish_message_to_group(message: Dict[str, Any], group: str) -> None:
        with current_app.producer_pool.acquire(block=True) as producer:
            producer.publish(
                msgpack.packb({
                  "__asgi_group__": group,
                  **message,
                }),
                exchange="groups",  # groups_exchange
                content_encoding="binary",
                routing_key=group,
                retry=False,  # Channel Layer at-most once semantics
            )

To call it, from a Celery-worker ``@task``...::

    publish_message_to_group({ "type": "chat.message", "text": "hi" }, "a-group")

... and a Django-Channels consumer like this will receive it::

    class WebsocketConnectionConsumer(AsyncWebsocketConsumer):
        async def connect(self):
            await self.channel_layer.group_add("a-group", self.channel_name)

        async def disconnect(self):
            await self.channel_layer.group_discard("a-group", self.channel_name)

        async def chat_message(self, event):
            assert event["text"] == "hi"

Alternatively, write your workers asynchronously, directly in `carehare`_.
It's more lightweight and faster than Celery, and the error handling is
simpler.

Design decisions
----------------

To scale enormously, this layer only creates one RabbitMQ queue per instance.
That means one web server gets one RabbitMQ queue, no matter how many
websocket connections are open. For each message being sent, the client-side
layer determines the RabbitMQ queue name and uses it as the routing key.

Groups are implemented using a single, global RabbitMQ direct exchange called
"groups" by default. To send a message to a group, the layer sends the message
to the "groups" exchange with the group name as the routing key. The client
binds and unbinds during ``group_add()`` and ``group_remove()`` to ensure
messages for any of its groups will reach it. See also the `groups_exchange`_
option.

RabbitMQ queues are ``exclusive``: when a client disconnects (through close or
crash), RabbitMQ will delete the queue and unbind the groups.

Once a connection has been created, it pollutes the event loop so that
``async_to_sync()`` will destroy the connection if it was created within
``async_to_sync()``. Each connection starts a background async loop that pulls
messages from RabbitMQ and routes them to receiver queues; each ``receive()``
queries receiver queues. Empty queues with no connections are deleted.

Deviations from the Channel Layer Specification
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The `Channel Layer Specification
<https://channels.readthedocs.io/en/latest/channel_layer_spec.html>`_ bends to
Redis-related restrictions. RabbitMQ cannot emulate Redis. Here are the
differences:

* **No ``flush`` extension**: To flush all state, simply disconnect all clients.
  (RabbitMQ won't allow one client to delete another client's data structures.)
* **No ``group_expiry`` option**: The `group_expiry option
  <https://channels.readthedocs.io/en/latest/channel_layer_spec.html#persistence>`_
  recovers when a ``group_add()`` has no matching ``group_discard()``. But the
  "group membership expiry" logic has a fatal flaw: it disconnects legitimate
  members. ``channels_rabbitmq`` addresses each root problem instead:

  * Web-server crash: RabbitMQ wipes all state related to a web server when
    the web server disconnects. There's no problem here for ``group_expiry``
    to solve.
  * Programming errors: You may err and call ``group_add()`` without
    eventually calling ``group_discard()``. Redis can't detect this
    programming error (because it can't detect web-server crashes). RabbitMQ
    can. The ``local_expiry`` option keeps your site running after you
    erroneously miss a ``group_discard()``. The channel layer warns when
    discarding expired messages. Monitor your server logs to detect your
    errors.
* **No "normal channels"**: `Normal channels
  <https://channels.readthedocs.io/en/latest/channel_layer_spec.html#channels>`_
  are job queues. In most projects, "normal channel" readers are worker
  processes, ideally divorced from Websockets and Django.

  If you want an async, RabbitMQ-based job queue, investigate `carehare
  <https://github.com/CJWorkbench/carehare>`_.

  If you're using Celery with the same RabbitMQ server, you can `publish
  messages from Celery`_, too.

Dependencies
------------

You'll need Python 3.8+ and a RabbitMQ server.

If you have Docker, here's how to start a development server::

    ssl/prepare-certs.sh  # Create SSL certificates used in tests
    docker run --rm -it \
         -p 5671:5671 \
         -p 5672:5672 \
         -p 15672:15672 \
         -v "/$(pwd)"/ssl:/ssl \
         -e RABBITMQ_SSL_CACERTFILE=/ssl/ca.cert \

CHANGELOG.rst
v4.0.1 - 2022-10-27
~~~~~~~~~~~~~~~~~~~

* Support Channels 4

v4.0.0 - 2021-06-08
~~~~~~~~~~~~~~~~~~~

* Try and forestall an error during reconnect. Connecting to RabbitMQ Node B
  after Node A went offline sometimes led to ``group_add()`` failures.
* Bump carehare to v1.0.0

**Upgrade instructions from v3.x**

* In accordance with RabbitMQ's URI specification, ``/`` is no longer allowed
  as the last character in an ``amqp`` URL. If your ``host`` ends with ``/``,
  delete the ``/``.

v3.0.0 - 2021-02-25
~~~~~~~~~~~~~~~~~~~

* Drop support for ``async_to_sync()`` without ``sync_to_async()``. If you have
  been using channels_rabbitmq to open a new RabbitMQ connection per
  message, you'll now get a RuntimeError.
  `Rationale <https://github.com/CJWorkbench/channels_rabbitmq/issues/28#issuecomment-734334065>`_
* Switched to `carehare <https://github.com/CJWorkbench/carehare>`_, so now we:
    * Expose errors. An error in the channel layers will be logged. (Previously,
      it could lead to a frozen layer.)
    * Release unawaited tasks and futures. Now, callers must await all futures.
      (Previously, there were missing warnings when callers didn't await.)
    * Expose ``get_channel_layer().carehare_connection``: a Future Connection.
      (It resets to a new Future upon disconnect.)
* Added ``await get_channel_layer().close()``, for clean shutdowns. On shutdown,
  all consumers will see ``StopConsuming``. (This is not in the Channel Layers
  Specification.)
* Stopped monkey-patching event-loop close. Now, ending the Python process
  without calling ``await get_channel_layer().close()`` may warn about unawaited
  coroutines. Use ASGI Lifespan to fix the warnings.

**Upgrade instructions from v2.x**

* Upgrade to **Python 3.8 or higher**.
* Ensure your ``host`` and ``ssl_context`` agree. If you use ``ssl_context``,
  your ``host`` must begin with ``amqps://``; otherwise, ``amqp://``.
* Ensure you do not call ``async_to_sync()`` outside of ``sync_to_async()``.

v2.0.0 - 2021-01-08
~~~~~~~~~~~~~~~~~~~

* ``local_capacity`` now counts messages. (Before, it was counting how many
  ``.get()`` calls were pending.)
* ``prefetch_count`` no longer exists. It was confusing when combined with
  ``local_capacity``.
* Switched to `aiormq <https://github.com/mosquito/aiormq>`_, which is
  better-maintained.
* Upgraded `msgpack <https://github.com/msgpack/msgpack-python>`_ to v1.

**Upgrade instructions from v1.x**

* If you have configured ``prefetch_count``, delete it.
* Upgrade to **Python 3.7 or higher**.
* Upgrade to **Django-Channels 3.0 or higher**.

v1.3.1 - 2020-12-29
~~~~~~~~~~~~~~~~~~~

* Fix package __version__ metadata, which prevented some users from
  installing v1.3.0. [issue #30] [PR #31]
* Fix async_to_sync() behavior on Python 3.6 and 3.7.

v1.3.0 - 2020-12-23
~~~~~~~~~~~~~~~~~~~

* Don't crash when called in async_to_sync() on asgiref >= 3.2.4
* Nix some deprecation warnings
* Allow installing alongside django-channels 3.0. [issue #29]

v1.2.1 - 2020-03-09
~~~~~~~~~~~~~~~~~~~

* Fix race when sending to group. [issue #23]

v1.2.0 - 2019-12-30
~~~~~~~~~~~~~~~~~~~

* Support Python 3.8 (upgrade to aioamqp=0.14.0) [issue #20] [PR #21]
* Deprecate ``group_expiry`` [issue #18] [PR #19]. Rationale:
  https://github.com/django/channels/issues/1371

Users who previously configured a ``group_expiry`` should remove it. It will
produce a deprecation warning, and it will be disallowed starting in v2.0.0.

v1.1.5 - 2019-08-22
~~~~~~~~~~~~~~~~~~~

* Allow asgiref=3.2.1. [issue #15]

v1.1.4 - 2019-07-09
~~~~~~~~~~~~~~~~~~~

* Avoid race which could drop messages and hang publish(). [issue #14]

v1.1.3 - 2019-06-16
~~~~~~~~~~~~~~~~~~~

* Warn when back-pressuring. Back-pressuring can make a web server
  unresponsive, even though it's behaving according to spec.
* Expire messages after receiving them from RabbitMQ but before
  ``receive()``. Warn when expiring. Allows recovery after messages
  are sent to nonexistent channels.

v1.1.2 - 2019-06-10
~~~~~~~~~~~~~~~~~~~

* Make ``groups_exchange`` configurable [issue #12]

v1.1.1 - 2019-05-23
~~~~~~~~~~~~~~~~~~~

* Retry after ``declare_queue`` errors during reconnect [issue #9]

v1.1.0 - 2019-05-09
~~~~~~~~~~~~~~~~~~~

* Add ``ssl_context`` configuration option [issue #2]
* Upgrade to aioamqp 0.13
* Avoid a 'Connection lost exc=None' warning when disconnecting from a thread
  with its own event loop [issue #4]

v1.0.1 - 2019-05-01
~~~~~~~~~~~~~~~~~~~

* Experimental support for Python 3.5. (There's no test coverage; use at your
  own risk!)

v1.0.0 - 2019-04-24
~~~~~~~~~~~~~~~~~~~

``channels_rabbitmq`` has been humming on production long enough that it
deserves the badge, "1.0".

``asgiref`` has seen a major version bump. ``channels_rabbitmq`` doesn't rely
on any of that changed code, but it does need newer dependencies so it can be
installed alongside the latest versions of ``channels`` and ``daphne``.

* Bump asgiref to 3.1 and channels to 2.2.

v0.0.11 - 2019-02-21
~~~~~~~~~~~~~~~~~~~~

* Bump msgpack from 0.5.2 to 0.6.1


channels_rabbitmq/core.py
from __future__ import annotations

import asyncio
import logging
import random
import string
import warnings
from typing import Iterable, Optional

import carehare
from channels.exceptions import ChannelFull, StopConsumer
from channels.layers import BaseChannelLayer

from .multiqueue import MultiQueue
from .reader import consume_into_multi_queue_until_connection_close
from .util import (
    ChannelRecipient,
    GroupRecipient,
    gather_without_leaking,
    serialize_message,
)

logger = logging.getLogger(__name__)

ReconnectDelay = 1.0  # seconds


async def _setup_connection(
    *,
    connection: carehare.Connection,
    queue_name: str,
    remote_capacity: Optional[int],
    expiry: Optional[int],
    groups_exchange: str,
) -> None:
    # Declare "groups" exchange. It may persist; spurious declarations
    # (such as on reconnect) are harmless.
    await connection.exchange_declare(groups_exchange, exchange_type="direct")

    # If we're reconnecting to RabbitMQ Node B after Node A disconnected us, our
    # previously-declared exclusive queue may still exist but _won't_ raise a
    # "resource-locked" error.
    #
    # [adamhooper, 2021-06-07] I'm not 100% sure this can actually happen. We
    # saw a problem that _looks_ like this on production, twice, when we
    # upgraded our node pools. The `queue_declare()` succeeded, but subsequent
    # `group_add()` operations failed because the queue did not exist. My theory
    # is: halfway along the way to deleting the exclusive queue, queue_declare()
    # succeeds, because maybe there's a race when deleting an exclusive queue in
    # which it appears non-exclusive to other nodes before it's deleted?
    #
    # ... this guess doesn't _literally_ violate the spec: "The client MAY NOT
    # attempt to use a queue that was declared as exclusive by another
    # still-open connection." Doesn't say anything about queues declared as
    # exclusive by another _closed_ connection on another node.
    #
    # If our woes persist, or if we go 1yr without seeing this message in
    # our logs, delete this block because we'll know it serves no purpose.
    # If, 1yr from now, we've solved our problem, let's build a reproduction
    # and report this upstream.
    try:
        await connection.queue_declare(queue_name, passive=True)
        # If we got here, the queue exists. The `queue_declare()` _below_ ought
        # to raise a 405 error if the queue still exists by the time we call it;
        # but we're guessing it might not, and there's no harm in raising the
        # same error ourselves.
        #
        # This exception will lead to a log message and a reconnect.
        #
        # TODO test using a whole RabbitMQ cluster, to expose this bug and
        # test this solution.
        raise carehare.ChannelClosedByServer(
            405, "channels_rabbitmq refusing to redeclare queue %s" % queue_name
        )
    except carehare.ChannelClosedByServer as err:
        if err.reply_code != 404:
            raise

    arguments = {}
    if remote_capacity is not None:
        arguments["x-max-length"] = remote_capacity
        arguments["x-overflow"] = "reject-publish"
    if expiry is not None:
        arguments["x-message-ttl"] = int(expiry * 1000)
    # This is an exclusive queue, because we want it to disappear when we
    # shut down. (Otherwise we'd leak queues every deploy.) And since an
    # exclusive queue disappears when we aren't connected, there's no way
    # to preserve every message across connects. That's okay -- disconnect
    # means one or two lost messages and nothing more.
    await connection.queue_declare(queue_name, exclusive=True, arguments=arguments)


async def _setup_subscriptions(
    *,
    connection: carehare.Connection,
    queue_name: str,
    groups_exchange: str,
    groups: Iterable[str],
) -> None:
    await gather_without_leaking(
        connection.queue_bind(
            queue_name=queue_name, exchange_name=groups_exchange, routing_key=group
        )
        for group in groups
    )


def _assert_channel_is_reply_channel(channel_name: str) -> None:
    if "!" not in channel_name:
        raise RuntimeError(
            """
***
You are trying to use "Normal Channels", as per
https://channels.readthedocs.io/en/1.x/concepts.html#channel-types.

channels_rabbitmq does not support Normal Channels. See
https://github.com/CJWorkbench/channels_rabbitmq/pull/11#issuecomment-499185070
for an explanation. You can build a pull request if you really want
this; but we heartily recommend you implement your work queue using
carehare directly. Your service will be simpler and it will scale better.
***"""
        )


def _random_letters(n: int) -> str:
    return "".join(random.choice(string.ascii_letters) for i in range(n))


class RabbitmqChannelLayer(BaseChannelLayer):
    """RabbitMQ channel layer.

    It routes all messages into a remote RabbitMQ server.

    Queue names look like "channels_asldkjfg"; Channel names look like
    "channels_asldkjfg!asdaSDGdlkgj". RabbitMQ only sees the queue name; the
    channel names are embedded in `message["__asgi_channel__"]`. Each
    Connection is responsible for a single queue and multiple channels on that
    queue. We call that queue a "channel_key".

    There are two "capacity" parameters: `remote_capacity` determines how many
    messages RabbitMQ will queue before a send raises `ChannelFull`.
    `local_capacity` determines the maximum number of messages to hold in
    memory in our ever-running pull-messages-from-RabbitMQ loop. If consumers
    are too slow then `local_capacity` messages will be buffered in memory; at
    that point Connection won't pull any more from RabbitMQ; so senders can
    send another `remote_capacity` messages to RabbitMQ and then they'll start
    raising `ChannelFull`.

    There is also an "expiry" parameter: this determines the minimum number of
    seconds a message must remain in RabbitMQ before being culled. A similar
    `local_expiry` determines the minimum amount of time a message must remain
    in the channels_rabbitmq buffers before it is acked (and potentially not
    delivered to all consumers).

    This layer does not implement "flush". To flush all state, simply
    disconnect all clients.
    """

    extensions = ["groups"]

    def __init__(
        self,
        host="amqp://guest:guest@127.0.0.1/asgi",
        local_capacity=100,
        remote_capacity=100,
        expiry=60,
        local_expiry: Optional[float] = None,
        group_expiry=None,
        ssl_context=None,
        groups_exchange="groups",
    ):
        self.host = host
        self.connect_timeout = 10.0  # seconds
        self.local_capacity = local_capacity
        self.remote_capacity = remote_capacity
        self.expiry = expiry
        self.local_expiry = local_expiry
        self.ssl_context = ssl_context
        self.groups_exchange = groups_exchange

        try:
            asyncio.get_running_loop()
        except RuntimeError:
            raise RuntimeError(
                """Refusing to initialize channel layer without a running event loop.

If you're writing a django-channels Worker, channels_rabbitmq does not support
this. Read the way to write workers in the carehare documentation.

If you're calling `async_to_sync()`, the call must be within code run by
`sync_to_async()`. Django Channels guarantees this for Django views. Elsewhere,
beware: `asgiref` *does* let you to call `async_to_sync()` without
`sync_to_async()`, but `channels_rabbitmq` *doesn't* -- hence this error."""
            )

        if group_expiry is not None:
            warnings.warn(
                (
                    "channels_rabbitmq does not support group_expiry. Please do not configure it. "
                    "For rationale, see "
                    "https://github.com/CJWorkbench/channels_rabbitmq/issues/18#issuecomment-547052373"
                ),
                category=DeprecationWarning,
            )

        # Choose queue name here: that way we can declare it on RabbitMQ with
        # exclusive=True and have it survive reconnections.
        self._queue_name = "channels_{rand}".format(rand=_random_letters(12))

        self._multi_queue = MultiQueue(capacity=local_capacity)
        self._carehare_connection: asyncio.Future[
            carehare.Connection
        ] = asyncio.Future()
        self._want_close: bool = False

    @property
    def carehare_connection(self) -> asyncio.Future[carehare.Connection]:
        if self._want_close:
            raise StopConsumer
        if self._carehare_connection.get_loop() != asyncio.get_running_loop():
            raise RuntimeError(
                "The caller tried using channels_rabbitmq on a different event loop than the one it was initialized with."
            )
        if not self._want_close and not hasattr(self, "_reconnect_forever_task"):
            self._reconnect_forever_task = asyncio.create_task(
                self._reconnect_forever()
            )
            if self.local_expiry is not None:
                self._expire_task = asyncio.create_task(
                    self._multi_queue.expire_locally_until_closed(
                        local_expiry=self.local_expiry
                    )
                )
        return self._carehare_connection

    async def _reconnect_forever(self):
        """Maintain self._carehare_connection.

        Cancel this task on event-loop shutdown (and at no other time). This
        will make `self.carehare_connection` raise `asyncio.CancelledError`.
        """
        EXPECTED_EXCEPTIONS = (
            carehare.ConnectionClosedByServer,
            carehare.ConnectionClosedByHeartbeatMonitor,
            OSError,
        )

        while not self._want_close:  # Reconnect (after disconnect) repeatedly
            while not self._want_close:  # Retry initial connect repeatedly
                try:
                    connection = carehare.Connection(
                        url=self.host,
                        connect_timeout=self.connect_timeout,
                        ssl=self.ssl_context,
                    )
                    await connection.connect()
                except (asyncio.TimeoutError, *EXPECTED_EXCEPTIONS) as err:
                    logger.exception(
                        "Failure connecting to RabbitMQ: %s. Retrying", str(err)
                    )
                    await asyncio.sleep(ReconnectDelay)
                    connection = None  # in case _want_close=True
                    continue  # retry

                try:
                    await _setup_connection(
                        connection=connection,
                        queue_name=self._queue_name,
                        remote_capacity=self.remote_capacity,
                        expiry=self.expiry,
                        groups_exchange=self.groups_exchange,
                    )
                    await _setup_subscriptions(
                        connection=connection,
                        queue_name=self._queue_name,
                        groups_exchange=self.groups_exchange,
                        groups=self._multi_queue._local_groups.keys(),
                    )
                except (carehare.ChannelClosedByServer, *EXPECTED_EXCEPTIONS) as err:
                    logger.exception(
                        "Failure declaring RabbitMQ objects: %s. Will retry.",
                        str(err),
                    )
                    await asyncio.sleep(ReconnectDelay)
                    connection = None  # in case _want_close=True
                    continue  # retry

                self._carehare_connection.set_result(connection)
                logger.info("Connected to RabbitMQ")
                break

            if self._want_close:
                break

            try:
                await consume_into_multi_queue_until_connection_close(
                    connection=connection,
                    channel=self._queue_name,
                    multi_queue=self._multi_queue,
                    prefetch_count=self.local_capacity,
                )
                await connection.closed
            except EXPECTED_EXCEPTIONS as err:
                logger.exception(
                    "Disconnected from RabbitMQ: %s. Will reconnect.", str(err)
                )

            self._carehare_connection = asyncio.Future()
            try:
                await connection.close()
            except EXPECTED_EXCEPTIONS:
                pass
            connection = None

        if connection is not None:
            try:
                await connection.close()
            except EXPECTED_EXCEPTIONS as err:
                logger.exception(
                    "Failure closing RabbitMQ connection: %s. Moving on....", str(err)
                )

    async def close(self):
        """Close our RabbitMQ connection and all tasks.

        This is not part of the Channel Layer Specification; but it should be a
        part of every server's lifecycle.

        After calling this, no more messages will be delivered. All pending and
        future calls to `receive()` will raise `StopConsumer` (causing all
        WebSockets clients to disconnect). Undelivered messages will be lost.

        After this method returns, the RabbitMQ connection will be closed and
        all internal tasks will be completed.
        """

        if not self._want_close:
            self._want_close = True
            self._multi_queue.close()

            if hasattr(self, "_expire_task"):
                self._expire_task.cancel()
            if (
                hasattr(self, "_reconnect_forever_task")
                and self._carehare_connection.done()
            ):
                await self._carehare_connection.result().close()

        if hasattr(self, "_expire_task"):
            try:
                await self._expire_task
            except asyncio.CancelledError:
                pass
            if hasattr(self, "_expire_task"):
                delattr(self, "_expire_task")

        if hasattr(self, "_reconnect_forever_task"):
            await self._reconnect_forever_task
            if hasattr(self, "_reconnect_forever_task"):
                delattr(self, "_reconnect_forever_task")

    async def send(self, channel, message):
        """Send a message onto a (general or specific) channel."""
        assert self.valid_channel_name(channel), "Channel name not valid"
        _assert_channel_is_reply_channel(channel)

        logger.debug("send %r to %s", message, channel)

        body = serialize_message(ChannelRecipient(channel), message)

        connection = await self.carehare_connection

        try:
            await connection.publish(
                body, exchange_name="", routing_key=channel.split("!")[0]
            )
        except carehare.ServerSentNack:
            raise ChannelFull from None
        except carehare.ConnectionClosed:
            logger.warning(
                "Failed to send message %r to %s because we disconnected",
                message,
                channel,
            )

    async def receive(self, channel):
        """Receive the first message that arrives on the channel.

        If more than one coroutine waits on the same channel, only one waiter
        will receive the message when it arrives.
        """
        # Make sure the channel name is valid then get the non-local part
        # and thus its index
        assert self.valid_channel_name(channel)
        _assert_channel_is_reply_channel(channel)
        assert channel.startswith(self._queue_name + "!")
        return await self._multi_queue.get(channel)

    async def new_channel(self, prefix=""):
        """Create a new channel name that can be used by something in our
        process as a specific channel.
        """
        return "%s!%s" % (self._queue_name, _random_letters(12))

    async def group_add(self, group, channel):
        """Add the channel name to a group.

        Spec deviation: `channel` must have been created on the same event loop
        as the RabbitMQ connection. In other words: you can't subscribe someone
        else's channel to a group.
        """
        assert self.valid_group_name(group), "Group name not valid"
        assert self.valid_channel_name(channel), "Channel name not valid"

        # Connect, to make sure the next line doesn't happen during _setup_subscriptions()
        connection = await self.carehare_connection

        # Now, assume _setup_subscriptions() isn't running, so there's no race
        # writing+reading self._multi_queue.
        n_bindings = self._multi_queue.group_add(group, channel)
        if n_bindings == 1:
            logger.debug("Binding queue %s to group %s", self._queue_name, group)
            # This group is new to our connection-level queue. Make a
            # connection-level binding.

            try:
                await connection.queue_bind(
                    queue_name=self._queue_name,
                    exchange_name=self.groups_exchange,
                    routing_key=group,
                )
            except carehare.ChannelClosedByServer as err:
                logger.exception(
                    "RabbitMQ refused to bind: %s. Messages to group %s will not reach clients on this server.",
                    str(err),
                    group,
                )
            except carehare.ConnectionClosed:
                pass  # we'll queue_bind() during reconnect

    async def group_discard(self, group, channel):
        """Remove the channel from the named group if it is in the group;
        does nothing otherwise (does not error)

        Spec deviation: `channel` must have been created on the same event loop
        as the RabbitMQ connection. In other words: you can't subscribe someone
        else's channel to a group.
        """
        assert self.valid_group_name(group), "Group name not valid"
        assert self.valid_channel_name(channel), "Channel name not valid"

        # Connect, to make sure the next line doesn't happen during _setup_subscriptions()
        connection = await self.carehare_connection

        # Now, assume _setup_subscriptions() isn't running, so there's no race
        # writing+reading self._multi_queue.
        n_bindings = self._multi_queue.group_discard(group, channel)
        if n_bindings == 0:
            logger.debug("Unbinding queue %s from group %s", self._queue_name, group)
            try:
                await connection.queue_unbind(
                    queue_name=self._queue_name,
                    exchange_name=self.groups_exchange,
                    routing_key=group,
                )
            except carehare.ChannelClosedByServer as err:
                logger.exception(
                    "RabbitMQ refused to unbind from group %s: %s. This suggests a bug somewhere.",
                    str(err),
                    group,
                )
            except carehare.ConnectionClosed:
                pass  # after reconnect, we're unbound anyway

    async def group_send(self, group, message):
        """Send a message to the entire group."""
        assert self.valid_group_name(group), "Group name not valid"

        logger.debug("group_send %r to %s", message, group)

        body = serialize_message(GroupRecipient(group), message)

        connection = await self.carehare_connection

        try:
            await connection.publish(
                body, exchange_name=self.groups_exchange, routing_key=group
            )
        except carehare.ServerSentNack:
            # "Sending to a group never raises ChannelFull; instead, it must
            # silently drop the message if it is over capacity, as per ASGI’s
            # at-most-once delivery policy."
            # https://channels.readthedocs.io/en/stable/channel_layer_spec.html#capacity
            #
            # ... let's at least _warn_....
            logger.warning("Aborting send to group %s: a queue is at capacity", group)
        except carehare.ConnectionClosed:
            logger.warning(
                "Failed to send message %r to group %s because we disconnected",
                message,
                group,
            )
